### DISCUSSION TITLE
How to use Ranger optimizer with YOLOV8?

### DISCUSSION THREAD

**Discussion by triciardvn** [ultralytics/ultralytics#16409] [Docs]
This is the error for my code since I want to try the "Ranger" optimizer instead of the existing available ones using YOLOV8.

`NotImplementedError: Optimizer 'Ranger' not found in list of available optimizers [Adam, AdamW, NAdam, RAdam, RMSProp, SGD, auto].To request support for addition optimizers please visit https://github.com/ultralytics/ultralytics.`

How can I integrate this new optimizer as I want to do an object detection for etrikes and ebikes. I already tried this code:

`!pip install git+https://github.com/lessw2020/Ranger21.git

from ultralytics import YOLO
import os

# Load your model
model = YOLO('yolov8n.pt')

# Define hyperparameters
hyperparameters = {
    'lr': 0.001,
    'weight_decay': 0.0005,
    'momentum': 0.937
}

# Get the number of training images
num_train_images = len(os.listdir('/content/drive/MyDrive/e/train/images'))
batch_size = 16
num_batches_per_epoch = num_train_images // batch_size
num_epochs = 10

# Setup training
model.train(data='/content/drive/MyDrive/e/data.yaml', epochs=num_epochs, imgsz=640, batch=batch_size, optimizer='Ranger')`

Please let me know what to do ASAP. Thank you.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @triciardvn, thank you for reaching out and for your interest in Ultralytics 🚀!

We understand you are eager to use the Ranger optimizer with YOLOv8. Unfortunately, it looks like the built-in support for Ranger isn't available yet. Your error stems from the optimizer not being recognized among the default options.

For issues like this, we recommend checking the [Docs](https://docs.ultralytics.com) where you might find related [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. This might inspire potential workarounds or help avoid common pitfalls.

If this is a 🐛 Bug Report or you're looking for guidance on custom implementation, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) along with any logs or error messages to better assist our team in diagnosing the issue.

Meanwhile, an Ultralytics engineer will join the discussion soon to offer further assistance!

Join our community for more interactive support:
- For real-time chat, head to [Discord](https://ultralytics.com/discord) 🎧.
- For in-depth discussions, visit [Discourse](https://community.ultralytics.com).
- For topic-specific threads, check out our [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

Ensure you're on the latest version of the `ultralytics` package along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLOv8 can be run in these verified environments:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are currently passing. These tests ensure YOLOv8 operates correctly across various platforms.

Stay tuned for more updates! 😊

**Comment by glenn-jocher** [comment]
@triciardvn currently, YOLOv8 does not support the Ranger optimizer directly. You can modify the training script to integrate custom optimizers by extending the existing codebase. For detailed guidance, consider exploring the Ultralytics documentation or community forums.

**Reply by triciardvn** [reply]
May I know the step by step procedure on how to customize it in the existing codebase? I actually searched for similar solutions and I tried to this: https://github.com/ultralytics/ultralytics/issues/4780#issuecomment-1710362992

But still it is not working on my end.

**Reply by glenn-jocher** [reply]
@triciardvn to customize the optimizer in YOLOv8, you'll need to modify the training script to include the Ranger optimizer. This involves integrating the optimizer into the PyTorch training loop. For detailed steps, consider reviewing the Ultralytics documentation or community discussions for guidance.
